sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -697,16 +697,21 @@ object ColumnPruning extends Rule[LogicalPlan] {
  *    `GlobalLimit(LocalLimit)` pattern is also considered.
  */
 object CollapseProject extends Rule[LogicalPlan] {
+  // If number of leaf expressions exceed MAX_LEAF_SIZE, do not collapse to prevent driver oom
+  // due to a single large project.
+  private val MAX_LEAF_SIZE = 1000
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
     case p1 @ Project(_, p2: Project) =>
```

Comment:
Seems like you can simplify it like this?;
```
  private def hasTooManyExprs(exprs: Seq[Expression]): Boolean = {
    var numExprs = 0
    exprs.foreach { _.foreach { _ => numExprs += 1 } }
    numExprs > SQLConf.get.XXXX
  }

  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
    case p1 @ Project(_, p2: Project) if hasTooManyExprs(p2.projectList) => // skip
      p1

    case p1 @ Project(_, p2: Project) =>
```

Reply:
Sure, I can simplify the condition check logic, do you suggest to add a new SQLConf instead of the hard limit? And for the case statement there is already a condition check called 'haveCommonNonDeterministicOutput', so I put them together. Also the same for  'case p @ Project(_, agg: Aggregate)'

Reply:
Yea, if we add this logic, I think we need a conf for that.